File: App/Models/OnboardingAnswers.swift
import Foundation

struct OnboardingAnswerChoice: Codable, Equatable {
    let value: String
    let label: String
}

/// Every answer of the Nova funnel, kept whole in `raw_answers.nova`: the server's answer columns
/// only cover the V2 questions. Single and multiple choices keep their Nova value and label; a
/// "Diğer" choice carries the text the user wrote as its label.
struct OnboardingNovaAnswers: Codable, Equatable {
    var flow = "nova-v1"
    let name: String?
    let certificate: OnboardingAnswerChoice?
    let work: OnboardingAnswerChoice?
    let role: OnboardingAnswerChoice?
    let experience: OnboardingAnswerChoice?
    let sectors: [OnboardingAnswerChoice]
    let trainings: [OnboardingAnswerChoice]
    let approach: [OnboardingAnswerChoice]
    let inspections: Int
    let growth: [OnboardingAnswerChoice]
    let assist: [OnboardingAnswerChoice]
    /// Question ids the user skipped.
    let skipped: [String]
    /// The optional updates box on the mail signup page; nil when that page was not used.
    let marketingEmailOptIn: Bool?

    enum CodingKeys: String, CodingKey {
        case flow, name, certificate, work, role, experience, sectors, trainings, approach
        case inspections, growth, assist, skipped
        case marketingEmailOptIn = "marketing_email_opt_in"
    }
}

struct OnboardingAnswersDraft: Codable, Equatable {
    let onboardingVersion: String
    let certificateClass: OnboardingAnswerChoice?
    let hazardClasses: [OnboardingAnswerChoice]
    let professionalRole: OnboardingAnswerChoice?
    let safetyProfileID: String?
    let appLanguage: String
    let sectors: [OnboardingAnswerChoice]
    let auditFrequency: OnboardingAnswerChoice?
    let selectedPlan: OnboardingAnswerChoice?
    let capturedAt: String
    /// Nova funnel only; drafts saved before it existed decode it as nil.
    let nova: OnboardingNovaAnswers?

    init(
        onboardingVersion: String = "v2",
        certificateClass: OnboardingAnswerChoice?,
        hazardClasses: [OnboardingAnswerChoice],
        professionalRole: OnboardingAnswerChoice? = nil,
        safetyProfileID: String? = nil,
        appLanguage: String = RDLanguage.current.rawValue,
        sectors: [OnboardingAnswerChoice],
        auditFrequency: OnboardingAnswerChoice?,
        selectedPlan: OnboardingAnswerChoice?,
        capturedAt: String = ISO8601DateFormatter().string(from: Date()),
        nova: OnboardingNovaAnswers? = nil
    ) {
        self.onboardingVersion = onboardingVersion
        self.certificateClass = certificateClass
        self.hazardClasses = hazardClasses
        self.professionalRole = professionalRole
        self.safetyProfileID = safetyProfileID
        self.appLanguage = appLanguage
        self.sectors = sectors
        self.auditFrequency = auditFrequency
        self.selectedPlan = selectedPlan
        self.capturedAt = capturedAt
        self.nova = nova
    }

    var hasProfileAnswers: Bool {
        certificateClass != nil ||
            !hazardClasses.isEmpty ||
            professionalRole != nil ||
            safetyProfileID != nil ||
            !sectors.isEmpty ||
            auditFrequency != nil ||
            nova != nil
    }

    var rpcPayload: OnboardingAnswersRPCPayload {
        OnboardingAnswersRPCPayload(
            onboardingVersion: onboardingVersion,
            certificateClass: certificateClass?.value,
            hazardClasses: hazardClasses.map(\.value),
            sectors: sectors.map(\.value),
            auditFrequency: auditFrequency?.value,
            selectedPlan: selectedPlan?.value,
            rawAnswers: OnboardingAnswersRawPayload(
                capturedAt: capturedAt,
                certificateClass: certificateClass,
                hazardClasses: hazardClasses,
                professionalRole: professionalRole,
                safetyProfileID: safetyProfileID,
                appLanguage: appLanguage,
                sectors: sectors,
                auditFrequency: auditFrequency,
                selectedPlan: selectedPlan,
                nova: nova
            )
        )
    }
}

struct OnboardingAnswersRawPayload: Codable, Equatable {
    let capturedAt: String
    let certificateClass: OnboardingAnswerChoice?
    let hazardClasses: [OnboardingAnswerChoice]
    let professionalRole: OnboardingAnswerChoice?
    let safetyProfileID: String?
    let appLanguage: String
    let sectors: [OnboardingAnswerChoice]
    let auditFrequency: OnboardingAnswerChoice?
    let selectedPlan: OnboardingAnswerChoice?
    let nova: OnboardingNovaAnswers?

    enum CodingKeys: String, CodingKey {
        case capturedAt = "captured_at"
        case certificateClass = "certificate_class"
        case hazardClasses = "hazard_classes"
        case professionalRole = "professional_role"
        case safetyProfileID = "safety_profile_id"
        case appLanguage = "app_language"
        case sectors
        case auditFrequency = "audit_frequency"
        case selectedPlan = "selected_plan"
        case nova
    }
}

struct OnboardingAnswersRPCPayload: Encodable {
    let onboardingVersion: String
    let certificateClass: String?
    let hazardClasses: [String]
    let sectors: [String]
    let auditFrequency: String?
    let selectedPlan: String?
    let rawAnswers: OnboardingAnswersRawPayload

    enum CodingKeys: String, CodingKey {
        case onboardingVersion = "p_onboarding_version"
        case certificateClass = "p_certificate_class"
        case hazardClasses = "p_hazard_classes"
        case sectors = "p_sectors"
        case auditFrequency = "p_audit_frequency"
        case selectedPlan = "p_selected_plan"
        case rawAnswers = "p_raw_answers"
    }
}

File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/onboarding/OnboardingAnswers.kt
package com.riskdetectedan.core.data.onboarding

import com.riskdetectedan.core.common.RdClientMetadata
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Kotlin mirror of App/Models/OnboardingAnswers.swift's OnboardingAnswerChoice. */
@Serializable
data class OnboardingAnswerChoice(val value: String, val label: String)

/**
 * Kotlin mirror of App/Models/OnboardingAnswers.swift's OnboardingNovaAnswers: every answer of the
 * Nova funnel, kept whole in `raw_answers.nova`, because the server's answer columns only cover the
 * V2 questions. Choices keep their Nova value and label; a "Diğer" choice carries the text the user
 * wrote as its label. No defaults, so every field is sent.
 */
@Serializable
data class OnboardingNovaAnswers(
    val flow: String,
    val name: String?,
    val certificate: OnboardingAnswerChoice?,
    val work: OnboardingAnswerChoice?,
    val role: OnboardingAnswerChoice?,
    val experience: OnboardingAnswerChoice?,
    val sectors: List<OnboardingAnswerChoice>,
    val trainings: List<OnboardingAnswerChoice>,
    val approach: List<OnboardingAnswerChoice>,
    val inspections: Int,
    val growth: List<OnboardingAnswerChoice>,
    val assist: List<OnboardingAnswerChoice>,
    /** Question ids the user skipped. */
    val skipped: List<String>,
    /** The optional updates box on the mail signup page; null when that page was not used. */
    @SerialName("marketing_email_opt_in") val marketingEmailOptIn: Boolean?,
)

/**
 * Kotlin mirror of App/Models/OnboardingAnswers.swift's OnboardingAnswersDraft — same fields,
 * same defaults (onboardingVersion "v2"). Locale and safety-profile defaults are resolved from
 * the active Android locale so onboarding, analysis and reporting use one coherent contract.
 */
@Serializable
data class OnboardingAnswersDraft(
    val onboardingVersion: String = "v2",
    val certificateClass: OnboardingAnswerChoice? = null,
    val hazardClasses: List<OnboardingAnswerChoice> = emptyList(),
    val professionalRole: OnboardingAnswerChoice? = null,
    val safetyProfileId: String? = RdClientMetadata.SAFETY_PROFILE_ID,
    val appLanguage: String = RdClientMetadata.APP_LANGUAGE,
    val sectors: List<OnboardingAnswerChoice> = emptyList(),
    val auditFrequency: OnboardingAnswerChoice? = null,
    val selectedPlan: OnboardingAnswerChoice? = null,
    val capturedAt: String = java.time.Instant.now().toString(),
    /** Nova funnel only; drafts saved before it existed load it as null. */
    val nova: OnboardingNovaAnswers? = null,
) {
    val hasProfileAnswers: Boolean
        get() = certificateClass != null ||
            hazardClasses.isNotEmpty() ||
            professionalRole != null ||
            safetyProfileId != null ||
            sectors.isNotEmpty() ||
            auditFrequency != null ||
            nova != null
}

/** Maps 1:1 to upsert_onboarding_v2_answers' p_* RPC parameters — same names, same shape. */
@Serializable
data class OnboardingAnswersRpcPayload(
    @SerialName("p_onboarding_version") val onboardingVersion: String,
    @SerialName("p_certificate_class") val certificateClass: String?,
    @SerialName("p_hazard_classes") val hazardClasses: List<String>,
    @SerialName("p_sectors") val sectors: List<String>,
    @SerialName("p_audit_frequency") val auditFrequency: String?,
    @SerialName("p_selected_plan") val selectedPlan: String?,
    @SerialName("p_raw_answers") val rawAnswers: OnboardingAnswersRawPayload,
)

@Serializable
data class OnboardingAnswersRawPayload(
    @SerialName("captured_at") val capturedAt: String,
    @SerialName("certificate_class") val certificateClass: RawChoice?,
    @SerialName("hazard_classes") val hazardClasses: List<RawChoice>,
    @SerialName("professional_role") val professionalRole: RawChoice?,
    @SerialName("safety_profile_id") val safetyProfileId: String?,
    @SerialName("app_language") val appLanguage: String,
    val sectors: List<RawChoice>,
    @SerialName("audit_frequency") val auditFrequency: RawChoice?,
    @SerialName("selected_plan") val selectedPlan: RawChoice?,
    /** Left out of the JSON when null (not a Nova draft), so it never clears a stored one. */
    val nova: OnboardingNovaAnswers? = null,
)

@Serializable
data class RawChoice(val value: String, val label: String)

fun OnboardingAnswerChoice.toRaw() = RawChoice(value, label)

fun OnboardingAnswersDraft.toRpcPayload(): OnboardingAnswersRpcPayload = OnboardingAnswersRpcPayload(
    onboardingVersion = onboardingVersion,
    certificateClass = certificateClass?.value,
    hazardClasses = hazardClasses.map { it.value },
    sectors = sectors.map { it.value },
    auditFrequency = auditFrequency?.value,
    selectedPlan = selectedPlan?.value,
    rawAnswers = OnboardingAnswersRawPayload(
        capturedAt = capturedAt,
        certificateClass = certificateClass?.toRaw(),
        hazardClasses = hazardClasses.map { it.toRaw() },
        professionalRole = professionalRole?.toRaw(),
        safetyProfileId = safetyProfileId,
        appLanguage = appLanguage,
        sectors = sectors.map { it.toRaw() },
        auditFrequency = auditFrequency?.toRaw(),
        selectedPlan = selectedPlan?.toRaw(),
        nova = nova,
    ),
)
